src/Control/Category/Cartesian.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE MonoLocalBinds #-}
+module Control.Category.Cartesian where
+
+import           Control.Category           (id, (>>>))
+import           Data.Functor.Contravariant
+import           Data.Void
+import           Prelude                    hiding (id)
+
+import           Control.Category.Tensor
+
+
+class Symmetric cat t => Semicartesian cat t where
+  split :: cat a (a `t` a)
+
+  fork :: cat a x -> cat a y -> cat a (x `t` y)
+  fork f g = split >>> f # g
+
+  infixr 9 /\
+  (/\) :: cat a x -> cat a y -> cat a (x `t` y)
+  (/\) = fork
+
+class Symmetric cat t => Semicocartesian cat t where
```

Comment:
@copytypecat So I was talking to @solomon-b about this yesterday: what do you think of using `Semicartesian` with `Flip`ped categories as a substitute for `Semicocartesian` and so on?

Specifically:

```
merge :: Semicartesian (Op cat) t => cat (a `t` a) a
merge = _
```

Reply:
My thinking is that it could be nice to have it this way because
1 the `Op` instances automatically generate all of those anyway, so this way is strictly "more" 
2 you instantly expose all those nice named dual functions rather than rewriting them using op trickery each time you want to use them 
3 this is more subjective, but I like the symmetry and feel queezy favoring one structure over the other... e.g. "products and sums" (or limits and colimits in general) in the universal property sense imo deserve to both be mentioned as their own thing (and then discussed as duals) rather than there merely being one thing and the other, its dual, an after thought

Reply:
Hmm. Could you flesh out what you mean by:

> 1 the Op instances automatically generate all of those anyway, so this approach is strictly "more"

I'm not quite able to follow what the "strictly more" bit is referring to.

Above I suggest defining newtype-unwrapped aliases of the stuff given by the `Op` instances. Perhaps you mean this is more effort than having the separate class?

>  2 you treat all of the named dual functions symmetrically to how you would treat their forward versions

I'm curious as to the asymmetry you see here. From my perspective, both `Type` and `Type^op` are independent cartesian categories (with different product structures). So we simply give them their own `Cartesian` instances.

I would imagine that with the 2 classes we're writing here, we would write four instances: a `Cartesian` instance for `(->) (,) ()`, a `Cartesian` instance for `(<-) Either Void`, a `Cocartesian` instance for `(->) Either Void`, and a `Cocartesian` instance for `(<-) (,) ()`. Is that right? Or do we omit a pair of these?

The main issue I have with the approach here is that there's essentially a combinatorial explosion of duplication that occurs in stuff that we hang off `Cartesian` and `Cocartesian`. Any operations you define in terms of `Cartesian` won't work for `Cocartesian` (and thus must be redefined), and vice versa.

If we want to be consistent, we should also e.g. create separate classes for "monoidal" and "opmonoidal" functors, instead of relying on instantiating the latter as  a special case of the former involving an opposite category. This would entail a similar tradeoff: class constraints/instances/top level definitions/etc. that refer to `Monoidal` would no longer work for things that instantiate `OpMonoidal`, and vice versa, and so we would have to have two suites of definitions.

Reply:
> So we simply give them their own `Cartesian` instances.

Sorry, my train of thought sort of petered out here. AFAICS there is no asymmetry in the two definitions so far (they're mirror images, except in so far as Haskell privileges the `->` arrow and we must newtype to obtain `<-`).

But perhaps the asymmetry is in the very fact that we take `Cartesian` and drop `Cocartesian`, and not vice versa. I'm admittedly less sure about this, but I feel like this is a question of "frame of reference". One is equally expressive as the other, so it is arbitrary which one we choose, but from my limited experience with the proofs I've seen, one does not typically bother to assume _both_ frame of references for the same proof. This would cause you to end up proving essentially the same proposition twice, which is analogous to us having to inhabit a parallel suite of definitions for `Cartesian` and `Cocartesian`.

Reply:
> Hmm. Could you flesh out what you mean by:

1 the Op instances automatically generate all of those anyway, so this approach is strictly "more"

I'm not quite able to follow what the "strictly more" bit is referring to.

I deleted this from my comment, so I'm backing off from that statement aside from affirming that indeed I perceive this approach as "less effort"

Reply:
Ah, I'm sorry, I wasn't keeping up with the edits. There was a bit in there that said:

> 1 the Op instances automatically generate all of those anyway, so this approach is strictly "more"

Didn't realize you'd already clarified it, thanks.

Btw, maybe we should chat about this in matrix.

Reply:
> Then it might be (visually / for the reader) annoying to put Op stuff when we want to equip one single category with both.

I definitely agree that this is the case. Would you say that defining e.g. `type (<-) = Op (->)` mitigates this at all?

Reply:
what about this point?
> Also, suppose we had some future scenarios where we'd say combine cartesian and co-cartesian, e.g. in a bicartesian category (i.e. categories with matrix-like morphisms). Then it might be (visually / for the reader) annoying to put Op stuff when we want to equip one single category with both.

Reply:
Well don't we want to stay in the general constraint of `Category cat` rather than `(->)`?

Reply:
> Well don't we want to stay in the general constraint of Category cat rather than (->)?

@copytypecat In general, certainly. But if we're saying: "when we want to equip one single category with both" (e.g. `->` or `~>` or whatever the case may be), I'm assuming (perhaps somewhat unrealistically) that it's possible to come up with an punchy way to refer to the opposite category in each such instance. It's also possible I'm missing the point with the bicartesian example you gave.